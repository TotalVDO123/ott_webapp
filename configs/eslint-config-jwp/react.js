module.exports = {
  extends: [
    // Extend the base config
    './typescript',

    // Use recommended React rules
    'plugin:react/recommended',
  ],

  rules: {
    // Not needed in React 17
    'react/react-in-jsx-scope': 'off',
  },

  overrides: [
    {
      files: ['*.jsx', '*.tsx', '*.ts'],
      plugins: [
        // Enable linting React code
        'react',
        'react-hooks',
      ],
      rules: {
        // Help with Hooks syntax
        'react-hooks/rules-of-hooks': 'error',
        'react-hooks/exhaustive-deps': 'error',

        // Handled by Typescript
        'react/prop-types': 'off',

        // This rule causes too many false positives, e.g. with default exports or child render function
        'react/display-name': 'off',
      },
    },
  ],

  settings: {
    react: {
      pragma: 'React',
      version: '17',
    },
  },
  parserOptions: {
    ecmaFeatures: {
      jsx: true,
    },
  },
};
