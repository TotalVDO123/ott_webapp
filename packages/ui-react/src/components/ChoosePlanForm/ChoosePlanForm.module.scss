.title {
  margin-bottom: 16px;
  font-weight: var(--body-font-weight-bold);
  font-size: 24px;
}

.tabs {
  display: flex;
  justify-content: center;
  margin-bottom: 16px;
}

.groupedButton {
  text-transform: capitalize;
}

.offers {
  display: flex;
  justify-content: center;
  margin: 0 -4px 24px;
  padding-bottom: 8px;
  overflow-x: auto;
  column-gap: 12px;
}


