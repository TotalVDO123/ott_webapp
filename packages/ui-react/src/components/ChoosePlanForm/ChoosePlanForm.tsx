import React, { useState, useMemo, useLayoutEffect } from 'react';
import { useTranslation } from 'react-i18next';
import type { FormErrors } from '@jwp/ott-common/types/form';
import type { Offer, ChooseOfferFormData } from '@jwp/ott-common/types/checkout';
import { testId } from '@jwp/ott-common/src/utils/common';

import Button from '../Button/Button';
import ButtonGroup from '../Button/ButtonGroup';
import FormFeedback from '../FormFeedback/FormFeedback';
import DialogBackButton from '../DialogBackButton/DialogBackButton';
import LoadingOverlay from '../LoadingOverlay/LoadingOverlay';
import PriceBox from '../PriceBox/PriceBox';

import styles from './ChoosePlanForm.module.scss';

type OfferPeriod = 'month' | 'year';

type Props = {
  values: ChooseOfferFormData;
  errors: FormErrors<ChooseOfferFormData>;
  onChange: React.ChangeEventHandler<HTMLInputElement | HTMLTextAreaElement>;
  setValue: (key: keyof ChooseOfferFormData, value: string) => void;
  onSubmit: React.FormEventHandler<HTMLFormElement>;
  onBackButtonClickHandler?: () => void;
  offers: Offer[];
  submitting: boolean;
};

const ChoosePlanForm: React.FC<Props> = ({ values, errors, submitting, offers, onChange, setValue, onSubmit, onBackButtonClickHandler }: Props) => {
  const { t } = useTranslation('account');
  const { selectedOfferId } = values;

  const groupedOffers = useMemo(
    () => offers.reduce((acc, offer) => ({ ...acc, [offer.period]: [...(acc[offer.period as OfferPeriod] || []), offer] }), {} as Record<OfferPeriod, Offer[]>),
    [offers],
  );

  const [offerFilter, setOfferFilter] = useState<OfferPeriod>(() => Object.keys(groupedOffers)[0] as OfferPeriod);

  useLayoutEffect(() => {
    setValue('selectedOfferId', groupedOffers[offerFilter as OfferPeriod]?.[0].offerId || '');

    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [offerFilter]);

  return (
    <form onSubmit={onSubmit} data-testid={testId('choose-offer-form')} noValidate>
      {onBackButtonClickHandler ? <DialogBackButton onClick={onBackButtonClickHandler} /> : null}
      <h1 className={styles.title}>{t('choose_offer.title')}</h1>
      {errors.form ? <FormFeedback variant="error">{errors.form}</FormFeedback> : null}
      <div className={styles.tabs}>
        <ButtonGroup>
          {Object.keys(groupedOffers).map((period) => (
            <Button
              key={period}
              label={t(`periods.${period}`)}
              onClick={() => {
                setOfferFilter(period as OfferPeriod);
              }}
              active={offerFilter === period}
              className={styles.groupedButton}
              data-testid={testId(`offer-period-${period}`)}
            />
          ))}
        </ButtonGroup>
      </div>
      <div className={styles.offers}>
        {!offers.length ? (
          <p>{t('choose_offer.no_pricing_available')}</p>
        ) : (
          groupedOffers[offerFilter].map((offer) => (
            <PriceBox key={offer.offerId} offer={offer} selected={selectedOfferId === offer.offerId} onChange={onChange} />
          ))
        )}
      </div>
      {submitting && <LoadingOverlay transparentBackground inline />}
      <Button
        label={t('choose_offer.continue')}
        disabled={!selectedOfferId || submitting || !offers.length}
        variant="contained"
        color="primary"
        type="submit"
        fullWidth
      />
    </form>
  );
};
export default ChoosePlanForm;
