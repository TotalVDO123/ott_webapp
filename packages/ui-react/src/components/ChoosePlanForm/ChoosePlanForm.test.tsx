import React from 'react';
import { axe } from 'vitest-axe';
import { fireEvent, render } from '@testing-library/react';
import type { Offer } from '@jwp/ott-common/types/checkout';
import monthlyOffer from '@jwp/ott-testing/fixtures/monthlyOffer.json';
import yearlyOffer from '@jwp/ott-testing/fixtures/yearlyOffer.json';

import ChoosePlanForm from './ChoosePlanForm';

const svodOffers = [monthlyOffer, yearlyOffer] as unknown as Offer[];

describe('<PlansForm>', () => {
  test('renders and matches snapshot', () => {
    const { container } = render(
      <ChoosePlanForm
        values={{ selectedOfferId: 'S916977979_NL', selectedOfferType: 'svod' }}
        errors={{}}
        onChange={vi.fn()}
        setValue={vi.fn()}
        onSubmit={vi.fn()}
        submitting={false}
        offers={svodOffers}
      />,
    );

    expect(container).toMatchSnapshot();
  });

  test('checks the monthly plan price correctly', () => {
    const { getByTestId } = render(
      <ChoosePlanForm
        values={{ selectedOfferId: 'S916977979_NL', selectedOfferType: 'svod' }}
        errors={{}}
        onChange={vi.fn()}
        setValue={vi.fn()}
        onSubmit={vi.fn()}
        submitting={false}
        offers={svodOffers}
      />,
    );

    expect(getByTestId('S916977979_NL')).toBeChecked();
  });

  test('checks the yearly plan price correctly', () => {
    const { getByTestId } = render(
      <ChoosePlanForm
        values={{ selectedOfferId: 'S345569153_NL', selectedOfferType: 'svod' }}
        errors={{}}
        onChange={vi.fn()}
        setValue={vi.fn()}
        onSubmit={vi.fn()}
        submitting={false}
        offers={svodOffers}
      />,
    );

    fireEvent.click(getByTestId('offer-period-year'));

    expect(getByTestId('S345569153_NL')).toBeChecked();
  });

  test('calls the onChange callback when changing the offer', () => {
    const onChange = vi.fn();
    const { getByTestId } = render(
      <ChoosePlanForm
        values={{ selectedOfferId: 'S916977979_NL', selectedOfferType: 'svod' }}
        errors={{}}
        onChange={onChange}
        setValue={vi.fn()}
        onSubmit={vi.fn()}
        submitting={false}
        offers={svodOffers}
      />,
    );

    fireEvent.click(getByTestId('offer-period-year'));

    fireEvent.click(getByTestId('S345569153_NL'));

    expect(onChange).toHaveBeenCalled();
  });

  test('calls the onSubmit callback when submitting the form', () => {
    const onSubmit = vi.fn();
    const { getByTestId } = render(
      <ChoosePlanForm
        values={{ selectedOfferId: 'S916977979_NL', selectedOfferType: 'svod' }}
        errors={{}}
        onChange={vi.fn()}
        setValue={vi.fn()}
        onSubmit={onSubmit}
        submitting={false}
        offers={svodOffers}
      />,
    );

    fireEvent.submit(getByTestId('choose-offer-form'));

    expect(onSubmit).toBeCalled();
  });

  test('WCAG 2.2 (AA) compliant', async () => {
    const { container } = render(
      <ChoosePlanForm
        values={{ selectedOfferId: 'S916977979_NL', selectedOfferType: 'svod' }}
        errors={{}}
        onChange={vi.fn()}
        setValue={vi.fn()}
        onSubmit={vi.fn()}
        submitting={false}
        offers={svodOffers}
      />,
    );

    expect(await axe(container, { runOnly: ['wcag21a', 'wcag21aa', 'wcag22aa'] })).toHaveNoViolations();
  });
});
