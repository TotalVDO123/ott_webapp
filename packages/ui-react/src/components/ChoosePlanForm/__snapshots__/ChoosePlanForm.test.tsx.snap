// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<PlansForm> > renders and matches snapshot 1`] = `
<div>
  <form
    data-testid="choose-offer-form"
    novalidate=""
  >
    <h1
      class="_title_eb8471"
    >
      choose_offer.title
    </h1>
    <div
      class="_tabs_eb8471"
    >
      <div
        class="_buttonGroup_f8f296"
      >
        <button
          class="_button_f8f296 _default_f8f296 _outlined_f8f296 _active_f8f296"
          type="button"
        >
          <span>
            Periods.month
          </span>
        </button>
        <button
          class="_button_f8f296 _default_f8f296 _outlined_f8f296"
          type="button"
        >
          <span>
            Periods.year
          </span>
        </button>
      </div>
    </div>
    <div
      class="_offers_eb8471"
    >
      <div
        aria-labelledby="title-S916977979_NL"
        class="_offer_059dc4"
      >
        <input
          checked=""
          class="_radio_059dc4"
          data-testid="S916977979_NL"
          id="S916977979_NL"
          name="selectedOfferId"
          type="radio"
          value="S916977979_NL"
        />
        <div
          class="_label_059dc4"
        >
          <label
            for="S916977979_NL"
          >
            <h2
              class="_offerTitle_059dc4"
              id="title-S916977979_NL"
            >
              Monthly subscription (recurring) to JW OTT Webapp
            </h2>
            <span
              class="hidden"
            >
              .
            </span>
            <hr
              aria-hidden="true"
              class="_offerDivider_059dc4"
            />
            <ul
              class="_offerBenefits_059dc4"
            >
              <li>
                <svg
                  aria-hidden="true"
                  class="_icon_585b29"
                  viewBox="0 0 24 24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M12 2C6.5 2 2 6.5 2 12S6.5 22 12 22 22 17.5 22 12 17.5 2 12 2M10 17L5 12L6.41 10.59L10 14.17L17.59 6.58L19 8L10 17Z"
                  />
                </svg>
                choose_offer.benefits.first_periods_free
                <span
                  class="hidden"
                >
                  .
                </span>
              </li>
              <li>
                <svg
                  aria-hidden="true"
                  class="_icon_585b29"
                  viewBox="0 0 24 24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M12 2C6.5 2 2 6.5 2 12S6.5 22 12 22 22 17.5 22 12 17.5 2 12 2M10 17L5 12L6.41 10.59L10 14.17L17.59 6.58L19 8L10 17Z"
                  />
                </svg>
                choose_offer.benefits.cancel_anytime
                <span
                  class="hidden"
                >
                  .
                </span>
              </li>
              <li>
                <svg
                  aria-hidden="true"
                  class="_icon_585b29"
                  viewBox="0 0 24 24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M12 2C6.5 2 2 6.5 2 12S6.5 22 12 22 22 17.5 22 12 17.5 2 12 2M10 17L5 12L6.41 10.59L10 14.17L17.59 6.58L19 8L10 17Z"
                  />
                </svg>
                choose_offer.benefits.watch_on_all_devices
                <span
                  class="hidden"
                >
                  .
                </span>
              </li>
            </ul>
            <div
              class="_fill_059dc4"
            />
            <div
              class="_offerPrice_059dc4"
            >
              € 6,99
               
              <small>
                /
                periods.month
              </small>
            </div>
          </label>
        </div>
      </div>
    </div>
    <button
      aria-disabled="false"
      class="_button_f8f296 _primary_f8f296 _fullWidth_f8f296"
      type="submit"
    >
      <span>
        choose_offer.continue
      </span>
    </button>
  </form>
</div>
`;
